README.md
# coordgeompy

A simple coordinate geometry helper package. This package is developed for the UBC MDS DSCI 524 Collaborative Software Development course at the University of British Columbia.

## Overview

This starter package allows users to perform various geometric operations like calculate distance between two parallel lines, distance between two n dimensional vectors, intersection of lines in 3-Dimensional space and so on. Our motivation in creating this package was to allow users with minimal experience in python coding to be able to perform these geometric calculations easily.
## Installation

```bash
$ pip install git+https://github.com/UBC-MDS/coordgeompy.git@v1.0.0
```

## Features

There are four main functions planned for development as outlined below. Additional functions may be added in the future.

Function 1 `dist_pll_lines_2d`: This function allows a user to calculate the distance between two parallel lines. This is the distance between the points where a perpendicular line intersects between the two parallel lines. This function will find that distance (d).

Function 2 `get_distance`: This function allows a user to calculate the the distance between two n dimensional vectors. Possible metrics that can be used with this function includes: Euclidean, Manhattan, Chebyshev, or Minkowski

Function 3 `is_intersection_3d`: This function allows a user to determine whether two infinite lines intersect in 3-dimensional space. The function will return True or False based on the input arguments.

Function 4 `is_orthogonal`: This function allows a user to determine whether two infinite lines are perpendicular in n-dimensional space. The function will return True or False based on the input arguments.

While we are not really reinventing the wheel on coordinate geometry calculations with our package, we used this as an opportunity to gain some experience in understanding how these calculations function in python language. There are existing packages that execute similar functions. For example, [SymPy](https://www.sympy.org/en/index.html) is a Python library that contains comprehensive mathematical functions including intersection of lines, shortest distance between a point and a line etc. We also found a package [coordinate-geometry 1.0.02](https://pypi.org/project/coordinate-geometry/) that had similar functions to calculating distance between two points, a point and a line, area of a triangle etc. [Scipy's](https://docs.scipy.org/doc/scipy/reference/generated/scipy.spatial.distance.cdist.html) `scipy.spatial.distance.cdist` functions similarly to Function 2. 
## Usage

- TODO

## Documentation 

The official documentation is hosted on Read the Docs: (tbd)
## Contributing

Interested in contributing? Check out the contributing guidelines. Please note that this project is released with a Code of Conduct. By contributing to this project, you agree to abide by its terms.

## Contributors

We welcome and recognize all contributors. Names and GitHub @usernames listed below:

- Arlin Cherian: @arlincherian
- Nico Van Den Hooff: @nicovandenhooff
- Zheren Xu: @ZherenXu
- Jordan Casoli: @jcasoli

## License

`coordgeompy` was created by Jordan Casoli, Nico Van Den Hooff, Arlin Cherian and Zheren Xu. It is licensed under the terms of the [![License: MIT](https://img.shields.io/badge/License-MIT-yellow.svg)](https://opensource.org/licenses/MIT)

## Credits

`coordgeompy` was created with [`cookiecutter`](https://cookiecutter.readthedocs.io/en/latest/) and the `py-pkgs-cookiecutter` [template](https://github.com/py-pkgs/py-pkgs-cookiecutter).


CHANGELOG.md
# Changelog

<!--next-version-placeholder-->

## v0.1.0 (11/01/2022)

- First release of `coordgeompy`!

tests/test_CoordGeomPy.py
from CoordGeomPy.CoordGeomPy import is_intersection_3d, dist_pll_lines_2d, get_distance, is_orthogonal
import numpy as np
import pytest


def test_is_intersection_3d_base():
    """Base tests on correct input"""
    
    # not intersecting
    assert is_intersection_3d((1, 3, -1), (0, -2, 4), (2, 1, 4), (0, 3, -3)) == False

    # parallel lines
    assert is_intersection_3d((1, 1, 1), (0, 0, 0), (1, 1, 1), (1, 1, 1)) == False

    # "normal" intersection
    assert is_intersection_3d((1, 0, 0), (0, 0, 0), (0, 1, 0), (0, 0, 0)) == True
    assert is_intersection_3d((1, 0, 0), (0, 0, 0), (0, 1, 0), (1, 1, 0)) == True

def test_is_intersection_3d_float():
    """Test to make sure function rounds floating point numbers"""
    
    # not intersecting
    assert is_intersection_3d((1.1, 3.2, -1.3), (0.0, -2.4, 4.2), (2.1, 1, 4), (0.2, 3.3, -3.0)) == False

    # parallel lines
    assert is_intersection_3d((1.1, 0.8, 0.9), (0, 0, 0), (1.2, 0.8, 1.4), (1, 1, 1)) == False

    # "normal" intersection
    assert is_intersection_3d((1.2, 0.3, 0.4), (0, 0.1, 0.1), (0, 1, 0), (0, 0, 0)) == True
    assert is_intersection_3d((1.3, 0.3, 0), (0, 0, 0), (0, 0.9, 0), (0.7, 1.3, 0)) == True

def test_is_intersection_3d_input():
    """Error testing"""

    m1 = {1, 1, 1}
    m2 = (0, 0, 0)
    b1 = (0, 0, 0)
    b2 = (0, 0, 0)

    # Invalid m1 type
    with pytest.raises(TypeError): 
        is_intersection_3d(m1, b1, m2, b2)

    # Invalid b1 type
    m1 = (1, 1, 1)
    b1 = {0, 0, 0}
    with pytest.raises(TypeError): 
        is_intersection_3d(m1, b1, m2, b2)

    # Invalid m1 length
    m1 = (1, 1, 1, 1)
    b1 = (0, 0, 0)
    with pytest.raises(ValueError): 
        is_intersection_3d(m1, b1, m2, b2)

# ==============================================================================================


def test_dist_pll_lines_2d_int():
    """Tests the distance between two parallel lines with integer input"""
    m = 2
    b1 = 4
    b2 = -1
    
    assert dist_pll_lines_2d(m, b1, b2) == 2.23606797749979
    assert round(dist_pll_lines_2d(m, b1, b2), 2) == 2.24

def test_dist_pll_lines_2d_float():
    """Tests the distance between two parallel lines with integer input"""
    m = 3.0
    b1 = 4.5
    b2 = 2.5
    
    assert dist_pll_lines_2d(m, b1, b2) == 0.6324555320336759
    assert round(dist_pll_lines_2d(m, b1, b2), 2) == 0.63

def test_dist_pll_lines_2d_invalid():
    """Tests the distance between two parallel lines with invalid input"""
    m = 2
    b1 = 4
    b2 = -1
    
    # invalid m value
    with pytest.raises(TypeError):
        dist_pll_lines_2d("a", b1, b2)
    
    with pytest.raises(TypeError):
        dist_pll_lines_2d(m, "abc", b2)
        
    with pytest.raises(TypeError):
        dist_pll_lines_2d(m, b1 , "xyz")
   
# ==============================================================================================


def test_get_distance_int():
    """Tests the get_distance function with integer input"""
    x1 = [1, 2, 3, 4]
    x2 = [5, 6, 7, 8]

    assert get_distance(x1, x1) == 0
    assert get_distance(x1, x2, "Euclidean") == 8
    assert get_distance(x1, x2, "Manhattan") == 16
    assert get_distance(x1, x2, "Chebyshev") == 4
    assert round(get_distance(x1, x2, "Minkowski", 3), 2) == 6.350


def test_get_distance_float():
    """Tests the get_distance function with float input"""
    x1 = [9.6, -0.6, 7]
    x2 = [1.23, 5.7, -3]

    assert get_distance(x1, x1) == 0
    assert round(get_distance(x1, x2, "euclidean"), 2) == 14.48
    assert round(get_distance(x1, x2, "manhattan"), 2) == 24.67
    assert get_distance(x1, x2, "chebyshev") == 10
    assert round(get_distance(x1, x2, "minkowski", -5), 2) == 5.94


def test_get_distance_invalid():
    """Tests that get_distance raises the right exceptions for invalid input"""

    x1 = [1, 2, 3, 4]
    x2 = [5, 6, 7, 8]
    x3 = [0]

    # invalid distance metric
    with pytest.raises(ValueError):
        get_distance(x1, x2, "invalid")

    # invalid x1 value
    with pytest.raises(TypeError):
        get_distance("a", x2, "euclidean")

    # invalid x2 value
    with pytest.raises(TypeError):
        get_distance(x2, "b", "euclidean")

    # mismatched dimensions
    with pytest.raises(TypeError):
        get_distance(x1, x3, "euclidean")

    # invalid p value with Minokowski distance
    with pytest.raises(TypeError):
        get_distance(x1, x2, "minkowski", "1")

# ==============================================================================================

def test_is_orthogonal_input():
    # Invalid input type
    with pytest.raises(TypeError): 
        is_orthogonal({1,2,3}, (1,2,3))
    with pytest.raises(TypeError): 
        is_orthogonal([1,2,3], 12)
    with pytest.raises(TypeError): 
        is_orthogonal("Wrong", [1,2,3])

    # Invalid length
    with pytest.raises(ValueError): 
        is_orthogonal([1,2], [1,1,1])
    with pytest.raises(ValueError): 
        is_orthogonal([], [1,1,1])

def test_is_orthogonal_T():
    assert is_orthogonal((0, 0, 0), (1, 1, 1)) == True
    assert is_orthogonal((1, 0, 0), (0, 1, 0)) == True
    assert is_orthogonal((13, -1, 2, 0), (-1, 1, 7, 10)) == True
    assert is_orthogonal([0.2, -4.8], [12, 0.5]) == True
    assert is_orthogonal((0.2, -4.8), [12, 0.5]) == True

def test_is_orthogonal_F():
    assert is_orthogonal((0, 0, 1), (1, 1, 1)) == False
    assert is_orthogonal((1, 1, 0), (-1, 2, 0)) == False
    assert is_orthogonal((13, -1, 2, 0), (-1, 1, 8, 10)) == False
    assert is_orthogonal([0.2, -4.8], [12.000001, 0.5]) == False


tests/test_coordgeompy.py
from coordgeompy.coordgeompy import is_intersection_3d, dist_pll_lines_2d, get_distance, is_orthogonal
import numpy as np
import pytest


def test_is_intersection_3d_base():
    """Base tests on correct input"""
    
    # not intersecting
    assert is_intersection_3d((1, 3, -1), (0, -2, 4), (2, 1, 4), (0, 3, -3)) == False

    # parallel lines
    assert is_intersection_3d((1, 1, 1), (0, 0, 0), (1, 1, 1), (1, 1, 1)) == False

    # "normal" intersection
    assert is_intersection_3d((1, 0, 0), (0, 0, 0), (0, 1, 0), (0, 0, 0)) == True
    assert is_intersection_3d((1, 0, 0), (0, 0, 0), (0, 1, 0), (1, 1, 0)) == True

def test_is_intersection_3d_float():
    """Test to make sure function rounds floating point numbers"""
    
    # not intersecting
    assert is_intersection_3d((1.1, 3.2, -1.3), (0.0, -2.4, 4.2), (2.1, 1, 4), (0.2, 3.3, -3.0)) == False

    # parallel lines
    assert is_intersection_3d((1.1, 0.8, 0.9), (0, 0, 0), (1.2, 0.8, 1.4), (1, 1, 1)) == False

    # "normal" intersection
    assert is_intersection_3d((1.2, 0.3, 0.4), (0, 0.1, 0.1), (0, 1, 0), (0, 0, 0)) == True
    assert is_intersection_3d((1.3, 0.3, 0), (0, 0, 0), (0, 0.9, 0), (0.7, 1.3, 0)) == True

def test_is_intersection_3d_input():
    """Error testing"""

    m1 = {1, 1, 1}
    m2 = (0, 0, 0)
    b1 = (0, 0, 0)
    b2 = (0, 0, 0)

    # Invalid m1 type
    with pytest.raises(TypeError): 
        is_intersection_3d(m1, b1, m2, b2)

    # Invalid b1 type
    m1 = (1, 1, 1)
    b1 = {0, 0, 0}
    with pytest.raises(TypeError): 
        is_intersection_3d(m1, b1, m2, b2)

    # Invalid m1 length
    m1 = (1, 1, 1, 1)
    b1 = (0, 0, 0)
    with pytest.raises(ValueError): 
        is_intersection_3d(m1, b1, m2, b2)

# ==============================================================================================


def test_dist_pll_lines_2d_int():
    """Tests the distance between two parallel lines with integer input"""
    m = 2
    b1 = 4
    b2 = -1
    
    assert dist_pll_lines_2d(m, b1, b2) == 2.23606797749979
    assert round(dist_pll_lines_2d(m, b1, b2), 2) == 2.24

def test_dist_pll_lines_2d_float():
    """Tests the distance between two parallel lines with integer input"""
    m = 3.0
    b1 = 4.5
    b2 = 2.5
    
    assert dist_pll_lines_2d(m, b1, b2) == 0.6324555320336759
    assert round(dist_pll_lines_2d(m, b1, b2), 2) == 0.63

def test_dist_pll_lines_2d_invalid():
    """Tests the distance between two parallel lines with invalid input"""
    m = 2
    b1 = 4
    b2 = -1
    
    # invalid m value
    with pytest.raises(TypeError):
        dist_pll_lines_2d("a", b1, b2)
    
    with pytest.raises(TypeError):
        dist_pll_lines_2d(m, "abc", b2)
        
    with pytest.raises(TypeError):
        dist_pll_lines_2d(m, b1 , "xyz")
   
# ==============================================================================================


def test_get_distance_int():
    """Tests the get_distance function with integer input"""
    x1 = [1, 2, 3, 4]
    x2 = [5, 6, 7, 8]

    assert get_distance(x1, x1) == 0
    assert get_distance(x1, x2, "Euclidean") == 8
    assert get_distance(x1, x2, "Manhattan") == 16
    assert get_distance(x1, x2, "Chebyshev") == 4
    assert round(get_distance(x1, x2, "Minkowski", 3), 2) == 6.350


def test_get_distance_float():
    """Tests the get_distance function with float input"""
    x1 = [9.6, -0.6, 7]
    x2 = [1.23, 5.7, -3]

    assert get_distance(x1, x1) == 0
    assert round(get_distance(x1, x2, "euclidean"), 2) == 14.48
    assert round(get_distance(x1, x2, "manhattan"), 2) == 24.67
    assert get_distance(x1, x2, "chebyshev") == 10
    assert round(get_distance(x1, x2, "minkowski", -5), 2) == 5.94


def test_get_distance_invalid():
    """Tests that get_distance raises the right exceptions for invalid input"""

    x1 = [1, 2, 3, 4]
    x2 = [5, 6, 7, 8]
    x3 = [0]

    # invalid distance metric
    with pytest.raises(ValueError):
        get_distance(x1, x2, "invalid")

    # invalid x1 value
    with pytest.raises(TypeError):
        get_distance("a", x2, "euclidean")

    # invalid x2 value
    with pytest.raises(TypeError):
        get_distance(x2, "b", "euclidean")

    # mismatched dimensions
    with pytest.raises(TypeError):
        get_distance(x1, x3, "euclidean")

    # invalid p value with Minokowski distance
    with pytest.raises(TypeError):
        get_distance(x1, x2, "minkowski", "1")

# ==============================================================================================

def test_is_orthogonal_input():
    # Invalid input type
    with pytest.raises(TypeError): 
        is_orthogonal({1,2,3}, (1,2,3))
    with pytest.raises(TypeError): 
        is_orthogonal([1,2,3], 12)
    with pytest.raises(TypeError): 
        is_orthogonal("Wrong", [1,2,3])

    # Invalid length
    with pytest.raises(ValueError): 
        is_orthogonal([1,2], [1,1,1])
    with pytest.raises(ValueError): 
        is_orthogonal([], [1,1,1])

def test_is_orthogonal_T():
    assert is_orthogonal((0, 0, 0), (1, 1, 1)) == True
    assert is_orthogonal((1, 0, 0), (0, 1, 0)) == True
    assert is_orthogonal((13, -1, 2, 0), (-1, 1, 7, 10)) == True
    assert is_orthogonal([0.2, -4.8], [12, 0.5]) == True
    assert is_orthogonal((0.2, -4.8), [12, 0.5]) == True

def test_is_orthogonal_F():
    assert is_orthogonal((0, 0, 1), (1, 1, 1)) == False
    assert is_orthogonal((1, 1, 0), (-1, 2, 0)) == False
    assert is_orthogonal((13, -1, 2, 0), (-1, 1, 8, 10)) == False
    assert is_orthogonal([0.2, -4.8], [12.000001, 0.5]) == False


src/CoordGeomPy/CoordGeomPy.py
import numpy as np


def dist_pll_lines_2d(m, b1, b2):
    """Finding the distance between two parallel lines.

    The distance between two parallel lines is the distance between the points where a perpendicular line intersects.
    This function will find that distance (d). The parameters of the function can be obtained from the equation of a parallel
    line y = mx + b.

    Parameters
    ----------
    m  : float
        slope of two parallel lines which are the same, i.e., slope of
        line 1, m1 = -(1/m2)
    b1 : float
        intercept of line 1 where y = mx + b1
    b2 : float
        intercept of line 2 where y = mx + b2

    Returns
    -------
    float
        The distance d between two parallel lines.

    Examples
    --------
    >>> dist_pll_lines_2d(3.0, 4.5, 2.5)
    0.6324
    >>> dist_pll_lines_2d(-4, 11, 23)
    2.9104
    """
    # apply corresponding values from the two parallel line equations to
    # the equation below
    if not isinstance(m, (int, float)):
        raise TypeError("m must be an integer or float")

    if not isinstance(b1, (int, float)):
        raise TypeError("b1 must be an integer or float")

    if not isinstance(b2, (int, float)):
        raise TypeError("b2 must be an integer or float")


    d = abs(b2 - b1) / np.sqrt(m ** 2 + 1)
    return d


def get_distance(x1, x2, metric="Euclidean", p=None):
    """Calculates the distance between two n dimensional vectors.

    Possible metrics include: Euclidean, Manhattan, Chebyshev, or Minkowski.

    Parameters
    ----------
    x1 : list of int or float
        The first n dimensional vector.
    x2 : list of int or float
        The second n dimensional vector.
    metric : str, default="Euclidean"
        The distance metric, must be one of "Euclidean", "Manhattan", Chebyshev,
        or "Minkowski"
    p : int, default=None
        The order of Minkowski distance to calculate.  Only required if `metric` is
        set to "Minkowski".

    Returns
    -------
    float
        The relevant distance between the two vectors.

    Examples
    --------
    >>> x1 = [1, 2, 3, 4]
    >>> x2 = [5, 6, 7, 8]
    >>> get_distance(x1, x2, metric="Euclidean")
    8
    >>> get_distance(x1, x2, metric="Manhattan")
    16
    >>> get_distance(x1, x2, metric="Chebyshev")
    4
    >>> get_distance(x1, x2, metric="Minkowski", 3)
    6.3496
    """

    # cast to lower for flexibility
    metric = metric.lower()

    # verify distance metric
    if metric not in ["euclidean", "manhattan", "chebyshev", "minkowski"]:
        raise ValueError(f"Invalid distance metric: {metric}")

    # verify that vectors are passed as python lists
    if not isinstance(x1, list) or not isinstance(x2, list):
        raise TypeError("x1 and x2 must be Python lists")

    # verify dimensions of vectors match
    if len(x1) != len(x2):
        raise TypeError("x1 and x2 must be the same length")

    # verify p if minkowski distance is metric
    if metric.lower() == "minkowski":
        if not isinstance(p, (int, float)):
            raise TypeError("p must be int or float")

    # convert to numpy arrays
    x1 = np.array(x1)
    x2 = np.array(x2)

    # check if arrays are equal if so distance is 0
    if np.array_equal(x1, x2):
        return 0
    elif metric == "euclidean":
        return np.sqrt(np.sum((np.square(x1 - x2))))
    elif metric == "manhattan":
        return np.sum(np.abs(x1 - x2))
    elif metric == "chebyshev":
        return np.max(np.abs(x1 - x2))
    elif metric == "minkowski":
        return np.power(np.sum(np.power(np.abs(x1 - x2), p)), 1 / p)


def is_intersection_3d(m1, b1, m2, b2):
    """Determines whether two infinite lines intersect in 3-dimensional space.

    Note that if two parallel lines are provided, they will be considered as NOT intersecting.
    Also note that this function expects integer values for x, y, z coordinates. Values will be rounded
    if integer values are not provided.

    This algorithm uses the following idea to test for intersection: Two (non parallel) lines intersect
    in 3d space if and only if they are coplanar.

    Parameters
    ----------
    m1 : list or tuple of floats
        This list corresponds to a 3-dimensional vector ⟨𝑚𝑥1,𝑚𝑦1,𝑚𝑧1⟩ describing the
        direction vector (or slope) of line 2. List or tuple must be of length 3.
    b1 : list or tuple of floats
        Any point on line 1. This list corresponds to a point (x1, y1, z1) that lies on line 1.
        This point must lie on line 2.
    m2 : list or tuple of floats
        This list corresponds to a 3-dimensional vector ⟨𝑚𝑥2,𝑚𝑦2,𝑚𝑧2⟩ describing the
        direction vector (or slope) of line 2.
    b2 : list or tuple of floats
        Any point on line 2. This list corresponds to a point (x2, y2, z2) that lies on line 2.
        This point must lie on line 2.

    Returns
    -------
    bool
        True if there is intersection, False if not.

    Examples
    --------
    >>> m1 = (1, 0, 0)
    >>> m2 = (0, 1, 0)
    >>> b1 = (0, 0, 0)
    >>> b2 = (0, 0, 0)
    >>> is_intersection_3d(m1, b1, m2, b2)
    True

    >>> m3 = (1, 3, -1)
    >>> m4 = (2, 1, 4)
    >>> b3 = (0, -2, 4)
    >>> b4 = (0, 3, -3)
    >>> is_intersection_3d(m3, m4, b3, b4)
    False
    """

    # Validate input
    if not isinstance(m1, (list, tuple) or not isinstance(m2, (list, tuple))):
        raise TypeError(
            "Only lists or tuples are supported. Please provide m1 & m2 as lists or tuples"
        )

    if not isinstance(b1, (list, tuple)) or not isinstance(b1, (list, tuple)):
        raise TypeError(
            "Only lists or tuples are supported. Please provide b1 & b2 as lists or tuples."
        )

    if len(m1) != 3 or len(m2) != 3 or len(b1) != 3 or len(b2) != 3:
        raise ValueError("All tuples should of length 3. Please check your inputs")

    # Ensure all values are integers
    m1 = (round(m1[0]), round(m1[1]), round(m1[2]))
    m2 = (round(m2[0]), round(m2[1]), round(m2[2]))
    b1 = (round(b1[0]), round(b1[1]), round(b1[2]))
    b2 = (round(b2[0]), round(b2[1]), round(b2[2]))

    m1 = np.array(m1)
    m2 = np.array(m2)
    b1 = np.array(b1)
    b2 = np.array(b2)

    # Check if lines are parallel
    if (m1 == m2).all():
        return False

    # Lines intersect if and only if they lie on the same plane (and are not parallel).
    x = np.cross(m1, m2)
    disp = np.array(b2) - np.array(b1)
    if np.dot(x, disp) == 0:
        return True
    else:
        return False


def is_orthogonal(m1, m2):
    """Determines whether two infinite lines are perpendicular in n-dimensional space.
    Parameters
    ----------
    m1 : list or tuple of floats
        This list corresponds to a n-dimensional vector ⟨mx1, my1, mz1, ...⟩ describing the
        direction vector of line 1.
    m2 : list or tuple of floats
        This list corresponds to a n-dimensional vector ⟨mx2, my2, mz2, ...⟩ describing the
        direction vector of line 2. Demensions of line 1 and line 2 must be equal.
    Returns
    -------
    bool
        True if there is orthogonal, False if not.
    Examples
    --------
    >>> m1 = (1, 0)
    >>> m2 = (0, 1)
    >>> is_orthogonal(m1, m2)
    True
    >>> m3 = (0, 0, 1)
    >>> m4 = (1, 1, 1)
    >>> is_orthogonal(m3, m4)
    False
    """
    # Validate input
    if not isinstance(m1, (list, tuple) or not isinstance(m2, (list, tuple))):
        raise TypeError("m1 and m2 must be lists or tuples")

    if len(m1) < 1 or len(m2) < 1:
        raise ValueError("The dimension of inputs must be at least 2")

    # Checking if m1 and m2 have the same dimension
    if len(m1) != len(m2):
        raise ValueError("m1 and m2 must be the same length")

    line1 = np.array(m1)
    line2 = np.array(m2)
    if round(np.dot(line1, line2), 10) == 0:
        return True
    else:
        return False
